components/brave_rewards/browser/rewards_service_impl.cc

```diff
@@ -1310,4 +1335,195 @@ brave_rewards::PublisherBanner RewardsServiceImpl::GetPublisherBanner(const std:
   return new_banner;
 }
 
+void RewardsServiceImpl::OnDonate(const std::string& publisher_key, int amount, bool recurring) {
+  if (recurring) {
+    SaveRecurringDonation(publisher_key, amount);
+    return;
+  }
+
+  ledger::PublisherInfo publisher_info(
+    publisher_key,
+    ledger::PUBLISHER_MONTH::ANY,
+    -1);
+
+  ledger_->DoDirectDonation(publisher_info, amount, "BAT");
+}
+
+bool SaveContributionInfoOnFileTaskRunner(const brave_rewards::ContributionInfo info,
+  PublisherInfoDatabase* backend) {
+  if (backend && backend->InsertContributionInfo(info))
+    return true;
+
+  return false;
+}
+
+void RewardsServiceImpl::OnContributionInfoSaved(const ledger::PUBLISHER_CATEGORY category, bool success) {
+  if (success && category == ledger::PUBLISHER_CATEGORY::DIRECT_DONATION) {
+    TipsUpdated();
```

Comment:
I don't think we should be doing this in RewardsServiceImpl. If the UI needs tip info when the contribution info is saved then we should have an observer method for it and let the webui call TipsUpdated. We're basically putting UI logic in the RewardsServiceImpl by doing things like this and I don't think it's a good idea to do that. If you want a single observer per profile to call this then you can create a new one, but it should be part of the UI code, not part of the service code

Reply:
the RewardsServiceImpl should just do the basics and let observers handle anything else that needs to happen when events occur